devtools/server/actors/replay/connection.js

```diff
@@ -240,6 +281,322 @@ class Recording extends EventEmitter {
   }
 }
 
+const RecordingState = {
+  READY: 0,
+  STARTING: 1,
+  RECORDING: 2,
+  STOPPING: 3
+};
+
+const recordings = new Map();
+
+function getRecordingKey(browser) {
+  return browser.frameLoader;
+}
+
+function getRecordingBrowser(key) {
+  return key.ownerElement;
+}
+
+function setRecordingState(key, state) {
+  recordings.set(key, {
+    state
+  });
+
+  Services.obs.notifyObservers({
+    browser: getRecordingBrowser(key),
+    state
+  }, "recordreplay-recording-changed");
+}
+
+function getRecordingState(browser) {
+  const {state} = recordings.get(getRecordingKey(browser)) || {state: RecordingState.READY};
+
+  return state;
+}
+
+// If an action invalidates the key (like updateBrowserRemoteness), we need to
+// remap the state from the old key to the new key.
+function remapRecordingState(browser, key) {
+  const newKey = getRecordingKey(browser);
+
+  if (recordings.has(key)) {
+    const entry = recordings.get(key);
+    recordings.delete(key);
+    recordings.set(newKey, entry);
+  }
+
+  return newKey;
+}
+
+function isRecording(browser) {
+  const {state} = getRecordingState(getRecordingKey(browser));
+
+  return state === RecordingState.RECORDING || browser.hasAttribute(
+    "recordExecution"
+  ) || isRecordingAllTabs();
```

Comment:
That's probably true. I was being conservative by keeping it in case the local state was out of sync. That said, probably better to catch the bug if we're out of sync than allow this fallback to fire.